File: src/lib/api/writing-session.ts
import { invoke } from '@tauri-apps/api/core';

import type { WritingSession } from './types';

export const writingSessionApi = {
	create: (date: string, wordsStart: number) =>
		invoke<WritingSession>('create_writing_session', {
			request: { date, words_start: wordsStart },
		}),

	getAll: () => invoke<WritingSession[]>('get_writing_sessions'),

	getByDate: (date: string) =>
		invoke<WritingSession | null>('get_writing_session_by_date', { date }),

	update: (
		id: string,
		request: { words_end?: number; duration_minutes?: number; scenes_edited?: string }
	) => invoke<WritingSession>('update_writing_session', { id, request }),

	delete: (id: string) => invoke<void>('delete_writing_session', { id }),
};

File: src-tauri/src/commands/writing_session.rs
use crate::models::{CreateWritingSessionRequest, UpdateWritingSessionRequest, WritingSession};
use crate::AppState;
use tauri::State;

#[tauri::command]
pub fn create_writing_session(
    state: State<'_, AppState>,
    request: CreateWritingSessionRequest,
) -> Result<WritingSession, String> {
    let db = state.get_db()?;
    let db = db.as_ref().ok_or("No project open")?;
    db.create_writing_session(&request.date, request.words_start)
}

#[tauri::command]
pub fn get_writing_sessions(state: State<'_, AppState>) -> Result<Vec<WritingSession>, String> {
    let db = state.get_db()?;
    let db = db.as_ref().ok_or("No project open")?;
    db.get_writing_sessions()
}

#[tauri::command]
pub fn get_writing_session_by_date(
    state: State<'_, AppState>,
    date: String,
) -> Result<Option<WritingSession>, String> {
    let db = state.get_db()?;
    let db = db.as_ref().ok_or("No project open")?;
    db.get_writing_session_by_date(&date)
}

#[tauri::command]
pub fn update_writing_session(
    state: State<'_, AppState>,
    id: String,
    request: UpdateWritingSessionRequest,
) -> Result<WritingSession, String> {
    let db = state.get_db()?;
    let db = db.as_ref().ok_or("No project open")?;
    db.update_writing_session(&id, &request)
}

#[tauri::command]
pub fn delete_writing_session(state: State<'_, AppState>, id: String) -> Result<(), String> {
    let db = state.get_db()?;
    let db = db.as_ref().ok_or("No project open")?;
    db.delete_writing_session(&id)
}
